### PR TITLE
[Enhancement](mysql) support prepare statement completely

### LINKED ISSUE #33386: [Enhancement] support prepare statement completely
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

now, doris just support prepare statement for point query.
it need support prepare statement completely for all query

### Solution

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nextdreamblue**
## Proposed changes

Issue Number: close #33386

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by morrySnow** [comment]
could we impl prepare on Nereids directly?

**Comment by nextdreamblue** [comment]
run buildall

**Review by dataroaring — CHANGES_REQUESTED** [review]
Please add regression tests.

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run p0

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run buildall

**Comment by nextdreamblue** [comment]
run compile

**Comment by nextdreamblue** [comment]
run feut

File: regression-test/data/point_query_p0/test_point_query_cluster_key.out [hunk 1]
```diff
@@ -180,19 +180,14 @@
 1231	119291.110000000	ddd	laooq	\N	2020-01-01 12:36:38	\N	1022-01-01	\N	1.111	\N	[119181.111100000, 819019.119100000, null]
 
 -- !point_select --
-1237	120939.111300000	a    ddd	laooq	2030-01-02	2020-01-01 12:36:38	22.822	7022-01-01	false	90696620686827832.374	[1.100000000, 2.200000000, 3.300000000, 4.400000000, 5.500000000]	[]
```
Comment (eldenmoon) [inline regression-test/data/point_query_p0/test_point_query_cluster_key.out:183]: why modified these cases？

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 2]
```diff
@@ -2666,6 +2674,10 @@ private void sendStmtPrepareOK() throws IOException {
         serializer.writeInt2(numParams);
         // reserved_1
         serializer.writeInt1(0);
+        // warning_count
+        serializer.writeInt2(0);
+        // metadata follows
+        serializer.writeInt1(1);
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:2680]: these two fields is under some condition, you should not directly adding them

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 3]
```diff
@@ -2702,7 +2720,7 @@ private void sendFields(List<String> colNames, List<Type> types) throws IOExcept
         // send field one by one
         for (int i = 0; i < colNames.size(); ++i) {
             serializer.reset();
-            if (prepareStmt != null && isExecuteStmt) {
+            if (prepareStmt != null && isExecuteStmt && prepareStmt.isPointQuery()) {
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:2723]: sendFields should not use prepareStmt.isPointQuery() it has not relation to this function

File: fe/fe-core/src/main/java/org/apache/doris/planner/ResultSink.java [hunk 4]
```diff
@@ -36,6 +36,8 @@ public class ResultSink extends DataSink {
 
     private TResultSinkType resultSinkType = TResultSinkType.MYSQL_PROTOCAL;
 
+    private boolean isBinaryRow = false;
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/planner/ResultSink.java:39]: rename to `isBinaryMysqlRowFormat`

File: gensrc/thrift/DataSinks.thrift [hunk 5]
```diff
@@ -199,6 +199,7 @@ struct TResultSink {
     1: optional TResultSinkType type;
     2: optional TResultFileSinkOptions file_options; // deprecated
     3: optional TFetchOption fetch_option;
+    4: optional bool is_binary_row;
```
Comment (eldenmoon) [inline gensrc/thrift/DataSinks.thrift:202]: rename `is_binary_row` to `is_binary_mysql_row_format`

File: fe/fe-core/src/main/java/org/apache/doris/analysis/PrepareStmt.java [hunk 6]
```diff
@@ -163,13 +164,15 @@ public void analyze(Analyzer analyzer) throws UserException {
         if (inner instanceof SelectStmt) {
             // Use tmpAnalyzer since selectStmt will be reAnalyzed
             Analyzer tmpAnalyzer = new Analyzer(context.getEnv(), context);
+            tmpAnalyzer.setPrepareStmt(true);
             SelectStmt selectStmt = (SelectStmt) inner;
             inner.analyze(tmpAnalyzer);
-            if (!selectStmt.checkAndSetPointQuery()) {
-                throw new UserException("Only support prepare SelectStmt point query now");
+            isPointQuery = selectStmt.checkAndSetPointQuery();
+            if (isPointQuery) {
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/PrepareStmt.java:171]: statment should be these types `NONE, FULL_PREPARED, STATEMENT`, STATEMENT type should suite. no need to `setPrepareStmt`, use type directly

File: regression-test/data/compaction/test_compaction_uniq_keys_row_store.out [hunk 7]
```diff
@@ -47,3 +47,51 @@
 -- !point_select --
```
Comment (eldenmoon) [inline regression-test/data/compaction/test_compaction_uniq_keys_row_store.out:47]: please add none full prepared type sql cases, you could refer to this PR : https://github.com/apache/doris/pull/33807/files

File: fe/fe-core/src/main/java/org/apache/doris/qe/MysqlConnectProcessor.java [hunk 8]
```diff
@@ -116,7 +116,7 @@ private void handleExecute() {
         String stmtStr = "";
         try {
             // new_params_bind_flag
-            if ((int) packetBuf.get() != 0) {
+            if (packetBuf.hasRemaining() && (int) packetBuf.get() != 0) {
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/qe/MysqlConnectProcessor.java:119]: please add more comment to explain the changes to mysql protocol in FE

File: fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java [hunk 9]
```diff
@@ -2666,6 +2674,10 @@ private void sendStmtPrepareOK() throws IOException {
         serializer.writeInt2(numParams);
         // reserved_1
         serializer.writeInt1(0);
+        // warning_count
```
Comment (eldenmoon) [inline fe/fe-core/src/main/java/org/apache/doris/qe/StmtExecutor.java:2677]: please add more comment to explain the changes to mysql protocol in FE

**Comment by eldenmoon** [comment]
I think we should merge this PR or [33806](https://github.com/apache/doris/pull/33806/files)

**Pull request closed without merging**